account_setup.py

```diff
@@ -0,0 +1,348 @@
+
+
+def create_aad_user(credentials, tenant_id, **kwargs):
+    """
+        Create an AAD application and service principal
+        :param credentials: msrestazure.azure_active_directory.AdalAuthentication
+        :param tenant_id: str
+        :param **application_name: str
+    """
+    graph_rbac_client = GraphRbacManagementClient(
+        credentials,
+        tenant_id,
+        base_url=AZURE_PUBLIC_CLOUD.endpoints.active_directory_graph_resource_id
+    )
+    application_credential = uuid.uuid4()
+    try:
+        display_name = kwargs.get("application_name", DefaultSettings.application_name)
+        application = graph_rbac_client.applications.create(
+            parameters=ApplicationCreateParameters(
+                available_to_other_tenants=False,
+                identifier_uris=["http://{}.com".format(display_name)],
+                display_name=display_name,
+                password_credentials=[
+                    PasswordCredential(
+                        end_date=datetime(2299, 12, 31, 0, 0, 0, 0, tzinfo=timezone.utc),
+                        value=application_credential,
+                        key_id=uuid.uuid4()
+                    )
+                ]
+            )
+        )
+        service_principal = graph_rbac_client.service_principals.create(
+            ServicePrincipalCreateParameters(
+                app_id=application.app_id,
+                account_enabled=True
+            )
+        )
+    except GraphErrorException as e:
+        if e.inner_exception.code == "Request_BadRequest":
+            application = next(graph_rbac_client.applications.list(
+                filter="identifierUris/any(c:c eq 'http://{}.com')".format(display_name)))
+
+            confirmation_prompt = "Previously created application with name {} found. "\
+                                  "Would you like to use it? (y/n): ".format(application.display_name)
+            prompt_for_confirmation(confirmation_prompt, e, ValueError("Response not recognized. Please try again."))
+
+            service_principal = next(graph_rbac_client.service_principals.list(
+                filter="appId eq '{}'".format(application.app_id)))
+        else:
+            raise e
+
+    return application.app_id, service_principal.object_id, str(application_credential)
+
+
+def create_role_assignment(credentials, subscription_id, scope, principal_id):
+    """
+        Gives service principal contributor role authorization on scope
+        :param credentials: msrestazure.azure_active_directory.AdalAuthentication
+        :param subscription_id: str
+        :param scope: str
+        :param principal_id: str
+    """
+    authorization_client = AuthorizationManagementClient(credentials, subscription_id)
+    role_name = 'Contributor'
+    roles = list(authorization_client.role_definitions.list(
+        scope,
+        filter="roleName eq '{}'".format(role_name)
+    ))
+    contributor_role = roles[0]
+    for i in range(10):
```

Comment:
What are you doing here? Is this trying to assign the roles on a retry loop?

Reply:
For some reason, the GraphRbacManagementClient's service_principal.create() method does not return an AzureOperationPoller to wait on (ala most of the other clients) nor does it block until the object is created server side. It just returns an object which may or may not have been created server side yet.

So I need to wait to give permissions here until the service principal is actually created.